Add parent submission field for reply threads

askmeaboutlo0m opened the pull request:
This adds a new "parent submission" field to the default submission section, letting the user pick a submission's parent from a drop-down. After posting a submission, website services get a chance to react to this and manipulate the child's part however they feel is useful.

Currently, this is implemented for Bluesky, Mastodon and Pleroma, which will set the "reply to post URL" field to the source URL of the parent post, removing the necessity to manually copy those over.

Other uses could be implemented if someone feels like it, different websites have various flavors of post relationships that could be handled this way.

The only time the application acts on parent-child relationships is right after a submission has been posted. That means invalid parents that reference removed submissions cause no harm, they're presented to the user as being unassigned. That means we don't need to bother maintaining the validity of these fields when removing submissions. We also don't need to check for circular dependencies, since those circles will break themselves as soon as one submission out of them is posted.

Amendment half a year later: there's also now `{parent:…}` "shortcuts" that get replaced with links to the parent when that is posted, allowing cross-linking.

askmeaboutlo0m committed:
Add parent submission field for reply threads

This adds a new "parent submission" field to the default submission
section, letting the user pick a submission's parent from a drop-down.
After posting a submission, website services get a chance to react to
this and manipulate the child's part however they feel is useful.

Currently, this is implemented for Bluesky, Mastodon and Pleroma, which
will set the "reply to post URL" field to the source URL of the parent
post, removing the necessity to manually copy those over.

Other uses could be implemented if someone feels like it, different
websites have various flavors of post relationships that could be
handled this way.

The only time the application acts on parent-child relationships is
right after a submission has been posted. That means invalid parents
that reference removed submissions cause no harm, they're presented to
the user as being unassigned. That means we don't need to bother
maintaining the validity of these fields when removing submissions. We
also don't need to check for circular dependencies, since those circles
will break themselves as soon as one submission out of them is posted.

foxyoreos commented:
If I'm understanding this, this is a hook for child submissions, right? When the parent submission gets posted, the child submissions run the update hook?

Do the resulting section-URLs/ids get stored beyond the hooks if Bluesky/Mastodon/Pleroma aren't submission sections? Probably beyond scope for the initial pull request, but it would be interesting to get this working with some kind of generic {prev} shortcut that could be used in any website description.

askmeaboutlo0m commented:
Yes, this is correct. And no, they don't get stored, but something of that sort would be useful anyway.

foxyoreos commented:
I got a chance to cherry-pick this down onto my local build and play with it - didn't notice any bugs while I was using it or in posting. I like that if part of an upload fails, the parts that succeeded still set the reply URL - that makes it really easy to recover from a failed upload. Having the field get handled at submission time to get rid of the need for ref cleanup is really clever, there are a number of edge cases that seem to just kind of work because of that. I'm going to keep using it locally.

My opinion on this does not in *any* way matter X3 ..but I think this is really great, I'm excited about this feature <3

And I'm just thinking of how difficult it is right now for me to handle scheduled threads - how I can't schedule multiple comic pages in advance because after each one gets posted I have to go update the next submission in the list.. this will save me a *ton* of time as an artist.  

----

A small thing jumps out to me: The parent submission select field isn't sorted by the same order as the submission overview page(s), which makes it a little more cumbersome to find submissions - but more importantly means that if two submissions have the same name (for example if you've scheduled a piece for two sites at two separate times or have an alt-version for a particular site), there's no way to know which submission is which in the selector.

Sorting off of `submission.order` wouldn't be enough to fix that, because you'd want scheduled submissions to be in a logical place too (and probably sorted by time or something), and including both File and Notification submissions in the same list means their orders would conflict - I would guess maybe these should be split up into sections and then sorted per-section? But it is not a big deal at all, and I might have just not understood the sorting?

----

The bigger thing that I wonder about is: I can't help but think that this might be more useful (and potentially a little bit simpler code-wise) if `supportsParentId` and the overridable `updateChildPart` were dropped in favor of setting a per-section field that could be arbitrarily consumed during `postFileSubmission` and `postNotificationSubmission`.

There are a couple of reasons that spring-to-mind:
1. My immediate thought when playing with this is that I want to be able to set `{prev}` links in descriptions. And sure, there are some per-website ways to do that so a website might want to have custom handling for that shortcut, but if it did it would want to do that in its shortcut handling code. And even if it didn't and it just resolved to a link, there aren't many (if any) websites where I wouldn't want that feature; so right off the bat I'm very tempted in my local build to add a default `updateChildPart` that just set a URL per-website no matter what.

2. Having the handling happen in the posting step instead of as a separate update would allow for possibly more intuitive overriding of the behavior as a user: so in the example of Bluesky, during posting the site could treat the parent submission as a fallback URL only if the user hasn't supplied a `replyToUrl` themselves.

   I would probably make some use of that behavior occasionally. I don't always treat Mastodon and Bluesky threads the same way, and I can think of scenarios where I'd have two sites in the same submission and would want to by paw set a replyURL for one of them, but have the other one be inherited from the parent. And there are ways I can think of to do that with the current system, but they seem kind of cumbersome and they're all kind of lossy because the only time the field can be set is when the parent is posted - so if the previously set URL gets overridden at that point or doesn't get set because the user already has a value in the field.. it seems to just add complications that the user could trip over.

3. Having the updated field be something separate from the user-submitted options means there's less chance of the user having a file open while its parent is posting and accidentally overwriting something when they save - but this is admittedly a pretty minor concern I think?

-----

I really want to stress though:

A) it doesn't matter what I think about this, I have nothing to do with Postybirb and there's no way you haven't spent more time than me thinking about it, so I'm posting my thoughts but they are not thoughts anypony needs to care about, and I don't know all of the considerations that have gone into the current design, there are probably good reasons why it was designed the way it was designed. Me rambling does not mean that I know what I'm talking about. If I don't know what I'm talking about, you can just ignore this X3

B) the feature in its current form is already something that would be really useful, and I don't want to detract from that. Being able to set up scheduled threads on Mastodon/Bluesky is a big deal to me, I am going to be using this locally no matter what and I expect it to save me a lot of time with those platforms. So the last thing I want to do is add feedback that would hold something up or over-complicate things. This is immediately going to be saving me time managing threads. <3

askmeaboutlo0m commented:
Don't put too much credit to my design or implementation here, I just built what I needed to cover my use case. I also have the use case of wanting to put links into the description, but what I need is cross-linking from a Telegram post to all other sites, so a `{prev}` shortcut wouldn't help me in that regard.

Sorting submissions would make sense, it's just good enough for me without it, since I can just search for the submission I want. So I didn't spend time thinking on or implementing any particular sorting.

Stashing the information in a generic field also makes sense, but you'd probably only implement it if you actually used it for something. And adding such a generic field is something you can just do in the future when it is actually needed, it's not something that causes compatibility issues, so I don't see enough incentive to change the current implementation. At worst, it would end up implementing something that doesn't actually cover what's needed, in which case you'd have to change it again and potentially deal with migrating the data, rather than just adding a field where there isn't one.

Although I definitely wouldn't store anything outside of what the user can control, otherwise you just obstruct manual operation. Letting the user fiddle with that field by hand makes it much easier to correct mistakes. Including the one of a user accidentally clobbering a link by saving over it, since they can just paste it back in by hand.

askmeaboutlo0m committed:
Allow automatic insertion of parent source links

When a submission is posted, it will now look through all child
descriptions and replace instances of `{parent:WHAT}` with a source URL.
`WHAT` can either be an internal website name or, if that would be
ambiguous, an account ID. The shortcut description shows the user a list
of options they can copy from. The account list page also shows these
account ids now.

askmeaboutlo0m commented:
Added `{parent:…}` "shortcuts" to this feature. When the parent is posted, they are replaced in child descriptions by the literal source links.
